Flake support

edolstra opened the pull request:
###### Motivation for this change

This PR:

* Adds a `flake.nix` file to Nixpkgs, making Nixpkgs usable from other flakes.
* Adds flake support to `nixos-rebuild`, enabling hermetic NixOS builds.
* Adds flake support to `nixos-container`.

A typical NixOS system configuration flake looks like this:

```nix
{
  edition = 201909;

  outputs = { self, nixpkgs, dwarffs, hydra }: rec {

    nixosConfigurations.my-server = nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      modules =
        [ dwarffs.nixosModules.dwarffs
          nixpkgs.nixosModules.notDetected
          hydra.nixosModules.hydra
          { system.configurationRevision = self.rev;
            /* typical configuration.nix stuff follows */
          }
        ];
    };

  };
}
```
which can be installed by doing
```
# nixos-rebuild switch --flake /path/to/flake --config my-server
```
(`--config` defaults to the host name so it can usually be omitted).

The option `system.configurationRevision` tracks the revision of the top-level flake. This allows the entire configuration to be reproduced reliably. In a running system, it can be queried using `nixos-version`:
```sh
$ nixos-version --json | jq -r .configurationRevision
2c475bc41c13e7a9ae5c16b6c10d8d328cd07ee7
```

Note: `NIX_PATH` is unavailable in flake-based builds. In particular this means that
```nix
imports = [ <nixpkgs/nixos/modules/installer/scan/not-detected.nix> ];
```
in `hardware-configuration.nix` no longer works. Instead you have to use
```nix
modules = [ nixpkgs.nixosModules.notDetected ];
```

`nixos-container` also supports flakes. For example, the following command creates [a container that serves the NixOS homepage](https://github.com/NixOS/nixos-homepage/blob/506001aa51806aac4f3680e61d9c9bee3466ef81/flake.nix#L97-L111):
```
$ nixos-container create homepage --flake nixos-homepage 
$ nixos-container start homepage 
$ curl http://$(nixos-container show-ip homepage)
```

TODO: add flake support to `nixos-install`, add tests.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

edolstra committed:
Add flake.nix

edolstra committed:
flake.nix: Set system explicitly for now

Flakes now are evaluated in pure mode, so we can't rely on
currentSystem anymore.

edolstra committed:
Export a function for building NixOS system configurations

edolstra committed:
Add flake.lock

edolstra committed:
legacyPkgs -> legacyPackages

edolstra committed:
flake.nix: Add check

edolstra committed:
make-tarball.nix: Derive version from lastModified instead of revCount

GitHub flakes don't provide revCount.

edolstra committed:
flake.nix: provides -> outputs

edolstra committed:
flake.nix: Update epoch

edolstra committed:
Remove flake.lock

edolstra committed:
Add flake outputs for the manuals

edolstra committed:
Update flake interface

edolstra committed:
flake.nix: Remove obsolete name attribute

edolstra committed:
nixos-rebuild: Add --flake option

edolstra committed:
nixosSystem: Automatically set version suffix from flake

E.g.

  $ nixos-version
  19.03.20190913.af5eb77 (Koi)

edolstra committed:
Add option system.configurationRevision to record revision of top-level flake

edolstra committed:
nixos-version: Add --json flag and show system.configurationRevision

edolstra committed:
Export nixosModules.notDetected

flokli commented on flake.nix:
```diff
@@ -0,0 +1,50 @@
+
+  outputs = { self }:
+    let
+      pkgs = import ./. { system = "x86_64-linux"; };
```
does this mean we currently can't use the nixpkgs flake from any non-x86 system?

edolstra replied:
Not from the `nix` command line. You can call it from other flakes with a non-x86_64-linux system argument though.

matthewbauer replied:
Can't we use builtins.currentSystem?

jD91mZM2 replied:
Nope, not in pure evaluation mode.

```
$ nix eval "(builtins.currentSystem)"
error: attribute 'currentSystem' missing, at (string):1:2
$ nix eval --impure "(builtins.currentSystem)"
"x86_64-linux"
```

arianvp commented:
I myself had a slightly different take on `nixos-rebuild` in the presence of flakes which I am using here: https://github.com/arianvp/nixos-stuff/blob/master/deployments.nix#L15-L46 (Though I'm still implementing the flake part, but the idea of calling `lib.nixosSystem` inside your package definition is the same to what I am doing).  

The idea is that you ou can do:
```
nix run github:arianvp/nixos-stuff:my-server -c deploy [switch|boot|test] [target*]
```

or in two steps:
```
nix build github:arianvp/nixos-stuff:my-server
./result/deploy switch

# or deploy remotely
./result/deploy switch  root@arianvp.me
```

E.g. a nixos flake _returns_ a `nixos-rebuild`-like command that can be called instead of having `nixos-rebuild` call `nix build`.  Kind of an inversion of control thing.

Would such a thing be more ergonomic? It would also mean you don't need `nixos-rebuild` on your computer to be able to deploy a nixos derivation.   (Because nixos-rebuild doesn't seem to be exposed as a package in nixpkgs.  so I do this so I can deploy nixos servers from my non-nixos machines)

edolstra committed:
nixos-rebuild: Use /etc/nixos/flake.nix if it exists

edolstra committed:
nixos-container: Add flake support

E.g. to create a container that runs the NixOS homepage:

  $ nixos-container create homepage --flake nixos-homepage

And to upgrade it:

  $ nixos-container update homepage

edolstra commented:
@arianvp Well, you already don't need `nixos-rebuild` since you can build the top-level system derivation and then run its switch-to-configuration script, e.g.
```
$ nix build .:nixosConfigurations.vyr.config.system.build.toplevel
$ ./result/bin/switch-to-configuration test
```
which is just what `nixos-rebuild` does.

edolstra commented:
Update: added flake support to `nixos-container`.

arianvp commented:
Well and you have to  `nix-env  --set` it right? Would there be a way to do that in one step?

edolstra committed:
nixos-{rebuild,container}: Use flakeref#attrpath syntax

This makes them consistent with the 'nix' command line.

Mic92 commented on flake.nix:
```diff
@@ -0,0 +1,50 @@
+{
+  epoch = 201909;
```
The pr mentions `edition`. Is it `edition` or `epoch`?

matthewbauer commented on flake.nix:
```diff
@@ -0,0 +1,50 @@
+      legacyPackages = pkgs;
+
+      nixosModules = {
+        notDetected = ./nixos/modules/installer/scan/not-detected.nix;
```
Why include this?

jtojnar replied:
> Note: NIX_PATH is unavailable in flake-based builds.

---

Should we add other modules as well? For example, I use the following images on my server:

    <nixpkgs/nixos/modules/profiles/minimal.nix>
    <nixpkgs/nixos/modules/virtualisation/container-config.nix>
    <nixpkgs/nixos/modules/installer/cd-dvd/channel.nix>

edolstra committed:
epoch -> edition

edolstra committed:
nixos-rebuild: uri -> url

edolstra committed:
nixos-rebuild: Support -L flag

edolstra committed:
flake.nix: Remove packages, builders outputs

edolstra committed:
flake.nix: Support more systems

Mic92 commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -218,16 +227,47 @@ if [ -z "$_NIXOS_REBUILD_REEXEC" ]; then
+       flakeAttr="${BASH_REMATCH[2]}"
+    fi
+    if [[ -z $flakeAttr ]]; then
+        hostname=$(cat /proc/sys/kernel/hostname)
```
```suggestion
        read hostname < /proc/sys/kernel/hostname
```

arianvp replied:
doesn't hostname=$(</proc/sys/kernel/hostname) work too? Not sure what the benefits are of the 3 different ways to write this

Mic92 replied:
```
read hostname < /proc/sys/kernel/hostname
```

is the fastest option since it does not require a subshell and hence forking. See also https://matthewbauer.us/blog/avoid-subshells.html

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/building-nixos-system-with-nix-build-and-a-channel-specifier/4747/2

arianvp commented on flake.nix:
```diff
@@ -0,0 +1,48 @@
+    in
+    {
+      lib = lib // {
+        nixosSystem = { modules, ... } @ args:
```
This function also already exists in as `pkgs.nixos`  (Can't hyperlink to it as `all-packages.nix` is too large for the UI).
Does it make sense to keep it there?

```nix
  nixos =
    configuration:
      let
        c = import (pkgs.path + "/nixos/lib/eval-config.nix") {
              inherit (pkgs.stdenv.hostPlatform) system;
              modules =
                [(
                  { lib, ... }: {
                    config.nixpkgs.pkgs = lib.mkDefault pkgs;
                  }
                )] ++ (
                  if builtins.isList configuration
                  then configuration
                  else [configuration]
                );
            };
      in
        c.config.system.build // c;
```

nrdxp commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -222,16 +229,47 @@ if [ -z "$_NIXOS_REBUILD_REEXEC" ]; then
+
+# Resolve the flake.
+if [[ -n $flake ]]; then
+    flake=$(nix flake info --json -- "$flake" | jq -r .uri)
```
What is the point of this conditional statement? Flakes do not appear to have an `uri` attribute so it always returns null. Tried changing it to `url` but got a git error when trying to run `nixos-rebuild`. If I remove this conditional entirely, it works without issue.

edolstra committed:
nixos-rebuild: Avoid subshell reading hostname

Co-Authored-By: Jörg Thalheim <Mic92@users.noreply.github.com>

edolstra committed:
flake.nix: Handle missing rev attribute

Dirty trees no longer set 'rev', so we need to handle this.

edolstra committed:
nixos-rebuild: Propagate various flake lock file flags

And also --refresh and --no-net.

samueldr commented on nixos/modules/installer/tools/nixos-version.sh:
```diff
@@ -8,6 +8,11 @@ case "$1" in
+  --json)
+    cat <<EOF
+{"nixosVersion": "@version@", "nixpkgsRevision": "@revision@", "configurationRevision": "@configurationRevision@"}
+EOF
```
Couldn't we replace a more generic `@nixosVersionJSONPayload@` with a properly serialized JSON value, rather than embedding a couple ad-hoc properties in the flow of the script?

yorickvP commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -225,16 +245,47 @@ if [ -z "$_NIXOS_REBUILD_REEXEC" ]; then
+
 # Re-execute nixos-rebuild from the Nixpkgs tree.
-if [ -z "$_NIXOS_REBUILD_REEXEC" -a -n "$canRun" -a -z "$fast" ]; then
+# FIXME: get nixos-rebuild from $flake.
```
This is important for [my workflow](https://xkcd.com/1172/)

edolstra replied:
The re-exec is fundamentally broken anyway, since it tries to build `nixos-rebuild` with the old Nix. (The code to upgrade Nix happens later.) We could do `nix build $flake#nixosConfigurations.$flakeAttr.config.system.build.nixos-rebuild` but I don't think it's useful currently.

yorickvP commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -225,16 +245,47 @@ if [ -z "$_NIXOS_REBUILD_REEXEC" ]; then
+fi
+
 # Find configuration.nix and open editor instead of building.
+# FIXME: handle flakes
```
FIXME: handle flakes.

yorickvP commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -296,7 +347,8 @@ prebuiltNix() {
 remotePATH=
 
-if [ -n "$buildNix" ]; then
+# FIXME: get nix from the flake.
```
FIXME: get nix from the flake.

yorickvP commented on pkgs/tools/virtualization/nixos-container/nixos-container.pl:
```diff
@@ -326,19 +359,35 @@ sub safeRemoveTree {
-         (defined $configFile && $configFile ne "")) {
-        writeNixOSConfig $nixosConfigFile;
+    # Unless overriden on the command line, rebuild the flake recorded
+    # in the container config file. FIXME: read the container config
```
fixme: read the container config in a more sensible way

yorickvP commented on pkgs/tools/virtualization/nixos-container/nixos-container.pl:
```diff
@@ -22,13 +22,27 @@
        nixos-container status <container-name>
-       nixos-container update <container-name> [--config <string>] [--config-file <path>]
+       nixos-container update <container-name>
+         [--config <string>]
```
this makes it harder to see that these arguments belong to `nixos-container --update`

edolstra replied:
I think it's more readable this way, and it's also the style used for `nixos-container create`.

yorickvP reviewed (changes requested):
Would be happy to see this merged. However, I'm afraid that the FIXME's in here will never get fixed once this is merged. We depend on the nixos-rebuild re'execing currently, so that fixme would be a blocker for moving to flakes for us, personally.

yorickvP commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -358,15 +410,37 @@ fi
+        if [[ -z $flake ]]; then
+            pathToConfig="$(nixBuild '<nixpkgs/nixos>' -A vm -k "${extraBuildFlags[@]}")"
+        else
+            echo "TODO: not implemented" >&2
```
todo: not implemented

edolstra replied:
This should be implemented by people who care about remote support (I don't).

yorickvP replied:
Not supporting nixos vm builds will *ensure* limited flake adoption.

edolstra replied:
I haven't used build-vm in years.

yorickvP replied:
I use it almost all the time, to do integration tests of modules. It's fairly prominent in the documentation and one of nixos's selling points.

edolstra replied:
Sure, but `nixos-rebuild build-vm` is just a wrapper around `nix build`. It doesn't do anything you couldn't otherwise do.

worldofpeace replied:
I've used build-vm to test the many many PRs that I've reviewed merged and tested. It's a very very useful tool for me.

worldofpeace replied:
Do you mean people can just do?
```
nix run vm -f '<nixpkgs/nixos>' --arg configuration "$PWD/configuration.nix" -c run-nixos-vm

yorickvP commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -358,15 +410,37 @@ fi
+        if [[ -z $flake ]]; then
+            pathToConfig="$(nixBuild '<nixpkgs/nixos>' -A vmWithBootLoader -k "${extraBuildFlags[@]}")"
+        else
+            echo "TODO: not implemented" >&2
```
todo: not implemented

yorickvP reviewed (changes requested):
^

alyssais commented:
I feel very conflicted about making this change now, when not only have we still not decided on the format for flake.nix, but the RFC seems to have been stalled for months…

pasqui23 commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -225,16 +245,47 @@ if [ -z "$_NIXOS_REBUILD_REEXEC" ]; then
 # Find configuration.nix and open editor instead of building.
+# FIXME: handle flakes
 if [ "$action" = edit ]; then
     NIXOS_CONFIG=${NIXOS_CONFIG:-$(nix-instantiate --find-file nixos-config)}
```
```suggestion
    NIXOS_CONFIG=${flake:-${NIXOS_CONFIG:-$(nix-instantiate --find-file nixos-config)}}
```
to handle the flake case

edolstra committed:
nixos-rebuild: Make 'edit' work with with flakes

edolstra committed:
nixos-rebuild: Remove --keep-going flag

edolstra committed:
nixos-version --json: Use builtins.toJSON

edolstra committed:
nixos.revision: Use null instead of "master"

"master" is not a valid SHA-1 commit hash, and it's not even
necessarily the branch used. 'nixos-version --revision' now returns an
error if the commit hash is not known.

edolstra committed:
nixos-rebuild: Remove TODOs

edolstra commented:
> Well and you have to nix-env --set it right? Would there be a way to do that in one step?

Yes, you can do `nix build --profile /nix/var/nix/profiles/system $flake#nixosConfigurations.$hostname.config.system.build.toplevel` to do the build and the profile update in one step.

FRidh reviewed (commented):
I would like to know how we deal on the stable with changes to e.g. the Flake format. Can it be relied on being stable during the lifetime of 20.03, or is it consider experimental? Probably needs a release note either way, clarifying this.

edolstra committed:
lib/tests/misc.nix: Fix dependency on currentSystem

This doesn't work in pure mode.

edolstra committed:
lib/tests/misc.nix: Don't make a copy on Nixpkgs

edolstra committed:
Fix 'nix flake check'

edolstra committed:
Doh

edolstra committed:
flake.nix: Add note

edolstra commented:
I've added a note that it's experimental (so it shouldn't be relied on).

arianvp commented:
Nixpkgs.lib.nixosSystem already exists as pkgs.nixos
It's confusing to have both. Can we remove one of the two?

edolstra commented:
Ideally we would remove `pkgs.nixos`, since it's not a package, but I don't want to make incompatible changes since this is all experimental.

nixos-discourse commented:
This issue has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/pinning-configuration-nix-with-niv/49031/4
